xtext/org.icyphy.linguafranca/src/lib/core/federate.c

```diff
@@ -48,74 +48,220 @@ THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
+ * connections from remote federates. This function
+ * only handles the creation of the server socket.
+ * The reserved port for the server socket is then
+ * sent to the RTI by sending an ADDRESS_AD message 
+ * (@see rti.h). This function expects no response
+ * from the RTI.
+ * 
+ * 
+ * If a port is specified by the user, that will be used
+ * as the only possibility for the server. This function
+ * will fail if that port is not available. If a port is not
+ * specified, the STARTING_PORT (@see rti.h) will be used.
+ * The function will keep incrementing the port in this case 
+ * until the number of tries reaches PORT_RANGE_LIMIT.
+ * 
+ * @note This function is similar to create_server(...) in rti.c.
+ * However, it contains specific log messages for the peer to
+ * peer connections between federates. It also additionally 
+ * sends an address advertisement (ADDRESS_AD) message to the
+ * RTI informing it of the port.
+ * 
+ * @param specified_port The specified port by the user.
+ */
+void create_server(int specified_port) {
+    int port = specified_port;
+    if (specified_port == 0) {
+        // Use the default starting port.
+        port = STARTING_PORT;
+    }
+    DEBUG_PRINT("Federate %d attempting to create a socket server on port %d.", _lf_my_fed_id, port);
+    // Create an IPv4 socket for TCP (not UDP) communication over IP (0).
+    int socket_descriptor = socket(AF_INET, SOCK_STREAM, 0);
+    if (socket_descriptor < 0) {
+        error_print_and_exit("Federate %d failed to obtain a socket server.", _lf_my_fed_id);
+    }
+
+    // Server file descriptor.
+    struct sockaddr_in server_fd;
+    // Zero out the server address structure.
+    bzero((char *) &server_fd, sizeof(server_fd));
+
+    server_fd.sin_family = AF_INET;            // IPv4
+    server_fd.sin_addr.s_addr = INADDR_ANY;    // All interfaces, 0.0.0.0.
+    // Convert the port number from host byte order to network byte order.
+    server_fd.sin_port = htons(port);
+
+    int result = bind(
+            socket_descriptor,
+            (struct sockaddr *) &server_fd,
+            sizeof(server_fd));
+    // If the binding fails with this port and no particular port was specified
+    // in the LF program, then try the next few ports in sequence.
+    while (result != 0
+            && specified_port == 0
+            && port >= STARTING_PORT
+            && port <= STARTING_PORT + PORT_RANGE_LIMIT) {
+        printf("Federate %d failed to get port %d. Trying %d\n", _lf_my_fed_id, port, port + 1);
+        port++;
+        server_fd.sin_port = htons(port);
+        result = bind(
+                socket_descriptor,
+                (struct sockaddr *) &server_fd,
+                sizeof(server_fd));
+    }
+    if (result != 0) {
+        if (specified_port == 0) {
+            error_print_and_exit("ERROR on binding the socket for federate %d. Cannot find a usable port. Consider increasing PORT_RANGE_LIMIT in federate.c",
```

Comment:
Exceeding acceptable line width.

Reply:
This is a very good point. With the introduction of new `write_to_socket` and `read_from_socket` functions, the line widths will generally be too long. Is there a way to automatically handle/detect these situations (e.g., by at least producing a warning somewhere)?

Reply:
Consoles usually line wrap. I was merely commenting on the code exceeding 80 chars. If you want to build a long string literal you can just write `"first part " "second part"` in C.

Reply:
I think 80 chars might be too restrictive especially given the indentation of the code here. Maybe we should try to adhere to 125 characters (which is the GitHub's line length limit) instead?

Reply:
IDK, I've always stuck to 80 chars for as long as I remember. We should write an actual style guide and settle on something that is agreeable to most contributors. I mostly flagged this because some lines were actually cut off while I was reviewing...

Reply:
I like the idea of settling on 125 characters.  The 80 character limit dates back to punched cards, which had 80 columns. In Ptolemy, we used 80 characters so we could print out the code in two column format for code reviews. But we aren't printing code anymore and our screens can easily handle 125 characters.